# Annotations

## Description
Custom gene annotations were prepared for both the mouse mm10 and *Drosophila* dm6 genome assemblies.

## Citation
>Asterix/Gtsf1 links tRNAs and piRNA silencing of retrotransposons<BR>
>Jonathan J. Ipsaro, Paul A. O'Brien, Shibani Bhattacharya, Arthur G. Palmer 3rd, Leemor Joshua-Tor<BR>
>Cell Rep. 2021 Mar 30;34(13):108914. doi: 10.1016/j.celrep.2021.108914.<BR>
>PMID: 33789107

## Sources Compiled for Custom Annotations

### Mouse
- Gencode version M21 (1)
- miRBase release 22.1 (2)
- piRNA cluster DB (3)
- TEtranscripts (4)
- tRNA DB(5)
- UCSC rRNA annotations (6)

### Fly
- FlyBase (7)
- miRBase release 22.1 (2)
- piRNA cluster DB (3)
- Custom annotations provided by A. Haase for piRNAs
- TEtranscripts (4)
- tRNA DB, UCSC (6)

## References
1.	A. Frankish et al., GENCODE reference annotation for the human and mouse genomes. <I>Nucleic Acids Res.</I> 47, D766-D773 (2019).
2.	A. Kozomara, M. Birgaoanu, S. Griffiths-Jones, miRBase: from microRNA sequences to function. <I>Nucleic Acids Res.</I> 47, D155-D162 (2019).
3.	D. Rosenkranz, piRNA cluster database: a web resource for piRNA producing loci. <I>Nucleic Acids Res.</I> 44, D223-230 (2016).
4.	Y. Jin, O. H. Tam, E. Paniagua, M. Hammell, TEtranscripts: a package for including transposable elements in differential expression analysis of RNA-seq datasets. <I>Bioinformatics</I> 31, 3593-3599 (2015).
5.	P. P. Chan, T. M. Lowe, GtRNAdb 2.0: an expanded database of transfer RNA genes identified in complete and draft genomes. <I>Nucleic Acids Res.</I> 44, D184-189 (2016).
6.	W. J. Kent et al., The human genome browser at UCSC. <I>Genome Res.</I> 12, 996-1006 (2002).
7.	J. Thurmond et al., FlyBase 2.0: the next generation. <I>Nucleic Acids Res.</I> 47, D759-D765 (2019).
